networkx/algorithms/isomorphism/tree_isomorph.py

```diff
@@ -0,0 +1,48 @@
+"""
+Tree isomorphism functions.
+"""
+import networkx as nx
+from networkx.exception import NetworkXError
+
+__author__ = """\n""".join(['Balazs Kossovics (kossovics@gmail.com)',
+    'Jeffrey Finkelstein (jeffrey.finkelstein@gmail.com)'])
+
+__all__ = ['rooted_tree_is_isomorphic',
+           'tree_is_isomorphic']
+
+# based on http://crypto.cs.mcgill.ca/~crepeau/CS250/2004/HW5+.pdf
+#   and jfinkels' comment
+# usage:
+# T1 = nx.random_powerlaw_tree(10, tries=1000)
+# T2 = nx.random_powerlaw_tree(10, tries=1000)
+# print tree_is_isomorphic(T1, T2)
+
+def canonical_form(T, root, _from=None):
+    try:
+        children = filter(lambda node: node != _from, T[root])
+        if len(children) == 0:
+            return ()
+        return tuple(sorted(canonical_form(T, v, root) for v in children))
+    except KeyError:
+        raise NetworkXError("Graph {} contains no node {}".format(T, root))
+
+def rooted_tree_is_isomorphic(T1, T2, r1, r2):
+    if not nx.is_tree(T1):
+        raise NetworkXError("Graph {} is not a tree.".format(T1))
+
+    if not nx.is_tree(T2):
+        raise NetworkXError("Graph {} is not a tree.".format(T2))
+
+    return canonical_form(T1, r1) == canonical_form(T2, r2)
+
+
+def tree_is_isomorphic(T1, T2):
+    if not T2:
```

Comment:
Hm, I believe that two null graphs should be considered isomorphic. The definition of isomorphic for graphs with node sets *U* and *V* is "there exists a function *f* from *U* to *V* such that *f* is a bijection and *f* preserves edges". Since both *U* and *V* are empty, there is just one function from *U* to *V*, the [empty function][1], and [it is a bijection][2]. Furthermore, the edge preservation requirement is [vacuously true][3].

However, the behavior of the tree isomorphism function should be "undefined" for non-tree inputs (that is, anything can happen). I think that, as with the `bipartite` package, the responsibility is on the calling code to decide whether the given graphs are actually trees before calling this `is_isomorphic` function. In other words, a pre-condition for calling this function will be that each graph is a tree. Since an empty graph is not a tree, you can assume that you will always get non-empty graphs as input.

[1]: https://en.wikipedia.org/wiki/Empty_function
[2]: http://math.stackexchange.com/questions/562412/is-the-empty-function-always-a-bijection
[3]: https://en.wikipedia.org/wiki/Vacuous_truth